cub/block/radix_rank_sort_operations.cuh

```diff
@@ -130,16 +119,518 @@ struct ShiftDigitExtractor : BaseDigitExtractor<KeyT>
+          reinterpret_cast<bit_ordered_type &>(src));
+
+        const std::uint32_t mask = (1 << bits_to_copy) - 1;
+        dst = dst | (((ordered_src >> src_bit_start) & mask) << dst_bit_start);
+
+        num_bits -= bits_to_copy;
+        dst_bit_start += bits_to_copy;
+      }
+      src_bit_start = 0;
+    }
+  }
+};
+
+template <class DecomposerT, class T>
+__host__ __device__ void digit(DecomposerT decomposer,
+                               std::uint32_t &dst,
+                               T &src,
+                               std::uint32_t &dst_bit_start,
+                               std::uint32_t &src_bit_start,
+                               std::uint32_t &num_bits)
+{
+  detail::for_each_member(digit_f{dst, dst_bit_start, src_bit_start, num_bits}, decomposer, src);
+}
+
+template <class DecomposerT>
+struct custom_digit_extractor_t
+{
+  DecomposerT decomposer;
+  std::uint32_t bit_start;
+  std::uint32_t num_bits;
+
+  __host__ __device__ __forceinline__ custom_digit_extractor_t(DecomposerT decomposer,
+                                                               std::uint32_t bit_start,
+                                                               std::uint32_t num_bits)
+      : decomposer(decomposer)
+      , bit_start(bit_start)
+      , num_bits(num_bits)
+  {}
+
+  template <class T>
+  __host__ __device__ __forceinline__ std::uint32_t Digit(T &key) const
+  {
+    std::uint32_t result{};
+    std::uint32_t dst_bit_start{};
+    std::uint32_t src_bit_start = bit_start;
+    std::uint32_t bits_remaining{num_bits};
+    digit(decomposer, result, key, dst_bit_start, src_bit_start, bits_remaining);
+    return result;
+  }
+};
+
+struct custom_bit_conversion_policy_t
+{
+  template <class DecomposerT, class T>
+  static __host__ __device__ T to_bit_ordered(DecomposerT decomposer, T val)
+  {
+    detail::radix::to_bit_ordered(decomposer, val);
+    return val;
+  }
+
+  template <class DecomposerT, class T>
+  static __host__ __device__ T from_bit_ordered(DecomposerT decomposer, T val)
+  {
+    detail::radix::from_bit_ordered(decomposer, val);
+    return val;
+  }
+};
+
+struct custom_bit_inversion_policy_t
+{
+  template <class DecomposerT, class T>
+  static __host__ __device__ T inverse(DecomposerT decomposer, T val)
+  {
+    detail::radix::inverse(decomposer, val);
+    return val;
+  }
+};
+
+template <class T>
+struct traits_t<T, false /* is_fundamental */>
+{
+  using bit_ordered_type              = T;
+  using bit_ordered_conversion_policy = custom_bit_conversion_policy_t;
+  using bit_ordered_inversion_policy  = custom_bit_inversion_policy_t;
+
+  template <class FundamentalExtractorT, class DecomposerT>
+  using digit_extractor_t = custom_digit_extractor_t<DecomposerT>;
+
+  template <class DecomposerT>
+  static __host__ __device__ bit_ordered_type min_raw_binary_key(DecomposerT decomposer)
+  {
+    T val{};
+    detail::radix::min_raw_binary_key(decomposer, val);
+    return val;
```

Comment:
I really dislike that pattern here. I would rather write 
```suggestion
    return detail::radix::min_raw_binary_key<T>(decomposer);
```
And pass the type explicitly rather than constructing using and returning

Reply:
The pattern will not disappear if we fix this place, it'll just go down the stack. Since it's an implementation detail that's consistent with the other functions here, I don't see the reason to change it.